File: p312.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int compute(vector<int>& nums, int left, int right, int k, vector<vector<int>>& dp) {
        int result = 0;
        if(k > left)
            result += dp[left][k-1];
        if(k < right)
            result += dp[k+1][right];        
        int pop = nums[k];
        if(left > 0)
            pop *= nums[left-1];
        if(right < nums.size()-1)
            pop *= nums[right+1];
        result += pop;        
        return result;
    }
    int maxCoins(vector<int>& nums) {
        int n = nums.size();
        if(n == 0)
            return 0;
        vector<vector<int>> dp(n, vector<int>(n, 0));
        // extend the length of subarray by i
        for(int i = 0; i < n; ++i) {
            // extend the subarray starting at j
            for(int j = 0; j < n-i; ++j) {
                // consider all the possibilities for last balloon to burst
                int left = j;
                int right = j+i;
                int best = 0;
                for(int k = left; k <= right; ++k) {
                    best = max(best, compute(nums, left, right, k, dp));
                }
                dp[left][right] = best;
            }
        }
        return dp[0][n-1];
    }
};

File: p312.py
from functools import lru_cache
from typing import List

class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        n = len(nums)

        @lru_cache(None)
        def dfs(start, end):
            if start == end:
                return 0
            ans = float('-inf')
            left = nums[start-1] if start > 0 else 1
            right = nums[end] if end < n else 1
            if start+1 == end:
                return left*nums[start]*right
            for i in range(start, end):
                ans = max(ans, dfs(start, i) + left*nums[i]*right + dfs(i+1, end))
            return ans

        return dfs(0, n)
